File: 백준/Gold/1914. 하노이 탑/하노이 탑.py
import sys

dp = {}

def hanoi(n, start, end, via):
    if (n == 1):
        return f"{start} {end}\n"
    
    if (f"{n},{start},{end},{via}" in dp):
        return dp[f"{n},{start},{end},{via}"]
    
    res = ""
    res += hanoi(n-1, start, via, end)
    res += f"{start} {end}\n"
    res += hanoi(n-1, via, end, start)
    dp[f"{n},{start},{end},{via}"] = res
    return res


n = int(sys.stdin.readline())
sys.stdout.write(f"{(2**n)-1}\n")

if (n <= 20):
    sys.stdout.write(hanoi(n, 1, 3, 2))

File: 백준/Gold/1914. 하노이 탑/하노이 탑.java
import java.io.*;
import java.math.BigInteger;
import java.util.*;

public class Main {
    private static BufferedWriter bw;
    private static BufferedReader br;
    private static HashMap<String, String> dp;
    public static void main(String[] args) throws IOException {
        dp = new HashMap<>();
        br = new BufferedReader(new InputStreamReader(System.in));
        bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int n = Integer.parseInt(br.readLine());
        bw.write(String.format("%s\n",
                new BigInteger("2").pow(n).subtract(new BigInteger("1")))
        );

//        System.out.printf("%s\n", new BigInteger("2").pow(n).subtract(new BigInteger("1")));
        if (n <= 20) {
            bw.write(hanoi(n, 1, 2, 3));
        }

        br.close();
        bw.flush();
        bw.close();
    }

    private static String hanoi(int n, int from, int via, int to) throws IOException{
        if (n == 1) {
            return String.format("%d %d\n",from, to );
        }

        if (dp.containsKey(String.format("%d,%d,%d,%d", n, from, via, to)))
            return dp.get(String.format("%d,%d,%d,%d", n, from, via, to));

        String res = "";
        res += hanoi(n-1, from, to, via);
        res += String.format("%d %d\n",from, to );
        res +=hanoi(n-1, via, from, to);

        dp.put(String.format("%d,%d,%d,%d", n, from, via, to), res);
        return res;
    }
}
